src/Functions/formatDateTime.cpp

```diff
@@ -113,168 +170,397 @@ class FunctionFormatDateTimeImpl : public IFunction
     class Action
     {
     public:
-        using Func = void (*)(char *, Time, const DateLUTImpl &);
+        using Func = std::conditional_t<
```

Comment:
@rschu1ze  I noticed that raw function pointer is much faster than `std::function` because of extra cost of dynamic memory allocation and virtual function call, please refer to this artical: https://blog.demofox.org/2015/02/25/avoiding-the-performance-hazzards-of-stdfunction/ 

Using `std::function` will cause performance degradation for function `formatDateTime`, which I thought is unacceptable for you guys. But `std::function` is required for new adding function `formatDateTimeWithJodaPattern` to capture variables. This is the reason why I use raw function pointer for `formatDateTime` and `std::function` for `formatDateTimeWithJodaPattern`. 
Query: `SELECT count() FROM numbers(10000000) WHERE NOT ignore(toDateTime('2017-01-01 00:00:00') + number % 100000000 + rand() % 100000 AS t, formatDateTime(t, '%Y-%m-%d %H:%M:%S')) ` 
If we use `std::function`, it will cost 0.838 sec, otherwise 0.381 sec.

Reply:
Do you think keeping the code and explaining it in comment is OK?

Reply:
Performance is a valid argument to do things unconventionally. In that case, you may keep things as is + add a comment (basically your above explanation)